File: packages/python/src/daraja/utils/metrics.py

from prometheus_client import Counter, Gauge, Histogram


class MetricsCollector:
    def increment(self, metric: str, tags: dict[str, str] | None = None, value: int = 1) -> None:
        pass

    def gauge(self, metric: str, value: float, tags: dict[str, str] | None = None) -> None:
        pass

    def timing(self, metric: str, duration_ms: float, tags: dict[str, str] | None = None) -> None:
        pass

    def histogram(self, metric: str, value: float, tags: dict[str, str] | None = None) -> None:
        pass


class NoopMetricsCollector(MetricsCollector):
    pass


class PrometheusMetricsCollector(MetricsCollector):
    def __init__(self, prefix: str = "mpesa_", default_tags: dict[str, str] | None = None) -> None:
        self._prefix = prefix
        self._default_tags = default_tags or {}

        self._counters: dict[str, Counter] = {}
        self._gauges: dict[str, Gauge] = {}
        self._histograms: dict[str, Histogram] = {}

        def _c(name: str, help: str, label_names: list[str] | None = None) -> Counter:
            c = Counter(f"{prefix}{name}", help, label_names or [])
            self._counters[name] = c
            return c

        def _g(name: str, help: str, label_names: list[str] | None = None) -> Gauge:
            g = Gauge(f"{prefix}{name}", help, label_names or [])
            self._gauges[name] = g
            return g

        def _h(
            name: str,
            help: str,
            label_names: list[str] | None = None,
            buckets: list[float] | None = None,
        ) -> Histogram:
            h = Histogram(
                f"{prefix}{name}",
                help,
                label_names or [],
                buckets=buckets or [0.01, 0.05, 0.1, 0.5, 1, 2, 5, 10],
            )
            self._histograms[name] = h
            return h

        _c(
            "requests_total",
            "Total number of API requests",
            ["operation", "environment", "status_code"],
        )
        _c("requests_success_total", "Successful requests", ["operation", "environment"])
        _c("requests_failed_total", "Failed requests", ["operation", "environment", "error_code"])
        _h(
            "request_duration_seconds",
            "Request duration distribution",
            ["operation", "environment"],
        )
        _c("token_refreshes_total", "Total OAuth token refreshes", ["status"])
        _g(
            "circuit_breaker_state",
            "Circuit breaker state (0=closed, 1=open, 2=half-open)",
            ["service"],
        )
        _c("circuit_breaker_failures_total", "Total circuit breaker failures", ["service"])
        _c("retry_attempts_total", "Total retry attempts", ["operation"])
        _c("retry_success_total", "Successful retries", ["operation"])
        _c("errors_total", "Total errors", ["error_type", "operation"])
        _c("api_errors_total", "API-specific errors", ["error_code", "operation"])
        _c("network_errors_total", "Network errors", ["operation"])
        _c("webhook_processed_total", "Total webhooks processed", ["webhook_type"])
        _c("webhook_success_total", "Successful webhooks", ["webhook_type"])
        _c("webhook_failed_total", "Failed webhooks", ["webhook_type"])
        _c("webhook_retried_total", "Webhooks that were retried", ["webhook_type"])
        _c("webhook_dlq_total", "Webhooks moved to DLQ", ["webhook_type"])
        _g("webhook_dlq_items", "Current DLQ item count")

    def increment(self, metric: str, tags: dict[str, str] | None = None, value: int = 1) -> None:
        counter = self._counters.get(metric)
        if counter:
            merged = {**self._default_tags, **(tags or {})}
            counter.labels(**merged).inc(value)

    def gauge(self, metric: str, value: float, tags: dict[str, str] | None = None) -> None:
        g = self._gauges.get(metric)
        if g:
            merged = {**self._default_tags, **(tags or {})}
            g.labels(**merged).set(value)

    def timing(self, metric: str, duration_ms: float, tags: dict[str, str] | None = None) -> None:
        self.histogram(metric, duration_ms / 1000.0, tags)

    def histogram(self, metric: str, value: float, tags: dict[str, str] | None = None) -> None:
        h = self._histograms.get(metric)
        if h:
            merged = {**self._default_tags, **(tags or {})}
            h.labels(**merged).observe(value)

File: sdks/go/types/metrics.go
package types

import (
	"github.com/prometheus/client_golang/prometheus"
)

type MetricsCollector interface {
	Increment(metric string, tags map[string]string, value int)
	Gauge(metric string, value float64, tags map[string]string)
	Timing(metric string, durationMs float64, tags map[string]string)
	Histogram(metric string, value float64, tags map[string]string)
}

type NoopMetricsCollector struct{}

func (n *NoopMetricsCollector) Increment(_ string, _ map[string]string, _ int)     {}
func (n *NoopMetricsCollector) Gauge(_ string, _ float64, _ map[string]string)     {}
func (n *NoopMetricsCollector) Timing(_ string, _ float64, _ map[string]string)    {}
func (n *NoopMetricsCollector) Histogram(_ string, _ float64, _ map[string]string) {}

type PrometheusMetricsCollector struct {
	counters   map[string]*prometheus.CounterVec
	gauges     map[string]*prometheus.GaugeVec
	histograms map[string]*prometheus.HistogramVec
}

func NewPrometheusMetricsCollector(prefix string) *PrometheusMetricsCollector {
	if prefix == "" {
		prefix = "mpesa_"
	}
	pm := &PrometheusMetricsCollector{
		counters:   make(map[string]*prometheus.CounterVec),
		gauges:     make(map[string]*prometheus.GaugeVec),
		histograms: make(map[string]*prometheus.HistogramVec),
	}

	pm.addCounter("requests_total", "Total number of API requests", []string{"operation", "environment", "status_code"})
	pm.addCounter("requests_success_total", "Successful requests", []string{"operation", "environment"})
	pm.addCounter("requests_failed_total", "Failed requests", []string{"operation", "environment", "error_code"})
	pm.addHistogram("request_duration_seconds", "Request duration distribution", []string{"operation", "environment"}, prometheus.DefBuckets)
	pm.addCounter("token_refreshes_total", "Total OAuth token refreshes", []string{"status"})
	pm.addGauge("circuit_breaker_state", "Circuit breaker state (0=closed, 1=open, 2=half-open)", []string{"service"})
	pm.addCounter("circuit_breaker_failures_total", "Total circuit breaker failures", []string{"service"})
	pm.addCounter("retry_attempts_total", "Total retry attempts", []string{"operation"})
	pm.addCounter("retry_success_total", "Successful retries", []string{"operation"})
	pm.addCounter("errors_total", "Total errors", []string{"error_type", "operation"})
	pm.addCounter("api_errors_total", "API-specific errors", []string{"error_code", "operation"})
	pm.addCounter("network_errors_total", "Network errors", []string{"operation"})
	pm.addCounter("webhook_processed_total", "Total webhooks processed", []string{"webhook_type"})
	pm.addCounter("webhook_success_total", "Successful webhooks", []string{"webhook_type"})
	pm.addCounter("webhook_failed_total", "Failed webhooks", []string{"webhook_type"})
	pm.addCounter("webhook_retried_total", "Webhooks that were retried", []string{"webhook_type"})
	pm.addCounter("webhook_dlq_total", "Webhooks moved to DLQ", []string{"webhook_type"})
	pm.addGauge("webhook_dlq_items", "Current DLQ item count", nil)

	return pm
}

func (pm *PrometheusMetricsCollector) addCounter(name, help string, labels []string) {
	opts := prometheus.CounterOpts{Name: name, Help: help}
	pm.counters[name] = prometheus.NewCounterVec(opts, labels)
	prometheus.MustRegister(pm.counters[name])
}

func (pm *PrometheusMetricsCollector) addGauge(name, help string, labels []string) {
	opts := prometheus.GaugeOpts{Name: name, Help: help}
	pm.gauges[name] = prometheus.NewGaugeVec(opts, labels)
	prometheus.MustRegister(pm.gauges[name])
}

func (pm *PrometheusMetricsCollector) addHistogram(name, help string, labels []string, buckets []float64) {
	opts := prometheus.HistogramOpts{Name: name, Help: help, Buckets: buckets}
	pm.histograms[name] = prometheus.NewHistogramVec(opts, labels)
	prometheus.MustRegister(pm.histograms[name])
}

func (pm *PrometheusMetricsCollector) Increment(metric string, tags map[string]string, value int) {
	if c, ok := pm.counters[metric]; ok {
		c.With(tags).Add(float64(value))
	}
}

func (pm *PrometheusMetricsCollector) Gauge(metric string, value float64, tags map[string]string) {
	if g, ok := pm.gauges[metric]; ok {
		g.With(tags).Set(value)
	}
}

func (pm *PrometheusMetricsCollector) Timing(metric string, durationMs float64, tags map[string]string) {
	pm.Histogram(metric, durationMs/1000.0, tags)
}

func (pm *PrometheusMetricsCollector) Histogram(metric string, value float64, tags map[string]string) {
	if h, ok := pm.histograms[metric]; ok {
		h.With(tags).Observe(value)
	}
}
